glow/rate_limiter.go

```diff
@@ -0,0 +1,63 @@
+package glow
+
+// Rate limiter utility using a sliding window approach.
+
+import (
+	"sync"
+	"time"
+)
+
+type RateLimiter struct {
+	limit int           // Maximum number of requests allowed
+	rate  time.Duration // Request rate for max requests
+	reqs  []time.Time   // Request list
+	mu    sync.Mutex
+}
+
+// Creates a new RateLimiter with limit requests per rate allowed.
+func NewRateLimiter(limit int, rate time.Duration) *RateLimiter {
+	return &RateLimiter{
+		limit: limit,
+		rate:  rate,
+		reqs:  make([]time.Time, 0),
+	}
+}
+
+// Ask the rate limiter if a request is allowed.
+// Returns true if allowed, false otherwise.
+func (r *RateLimiter) Allow() bool {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	now := time.Now()
+
+	// expire requests
+	exp := now.Add(-r.rate)
+	idx := -1
+	for i, t := range r.reqs {
+		if t.After(exp) {
+			idx = i
+			break
+		}
+	}
+
+	if idx == -1 {
+		r.reqs = r.reqs[:0] // Clear the list since none of the entries were after the expiry time
+	} else {
+		r.reqs = r.reqs[idx:] // Retain starting at first index after the expiry time
+	}
+
+	if len(r.reqs) < r.limit {
+		r.reqs = append(r.reqs, now)
+		return true
+	}
+
+	return false
+}
+
+// Clear the rate limiter. Used for testing.
+func (r *RateLimiter) Clear() {
```

Comment:
I'm not a huge fan of having this function, I feel like it will lead to tests that are less representative of the real world operation of the server

Reply:
Not to have it makes testing harder, I will see if I can find a way to do it